client/session.go

```diff
@@ -1224,79 +1246,88 @@ func (s *session) FetchBlocksMetadataFromPeers(
 		close(metadataCh)
 		close(errCh)
 	}()
 
 	return newMetadataIter(metadataCh, errCh), nil
 }
 
+// FetchBootstrapBlocksFromPeers will fetch the specified blocks from peers for
+// bootstrapping purposes. Refer to peer_bootstrapping.md for more details.
 func (s *session) FetchBootstrapBlocksFromPeers(
 	nsMetadata namespace.Metadata,
 	shard uint32,
 	start, end time.Time,
 	opts result.Options,
+	version FetchBlocksMetadataEndpointVersion,
 ) (result.ShardResult, error) {
+	if version != FetchBlocksMetadataEndpointV1 && version != FetchBlocksMetadataEndpointV2 {
+		return nil, errInvalidFetchBlocksMetadataVersion
+	}
+
 	var (
 		result   = newBulkBlocksResult(s.opts, opts)
-		complete = int64(0)
-		doneCh   = make(chan error, 1)
-		onDone   = func(err error) {
-			atomic.StoreInt64(&complete, 1)
-			select {
-			case doneCh <- err:
-			default:
-			}
-		}
-		waitDone = func() error {
-			return <-doneCh
-		}
-		m = s.streamFromPeersMetricsForShard(shard, resultTypeBootstrap)
+		doneCh   = make(chan struct{})
+		progress = s.newPeerMetadataStreamingProgressMetrics(shard, resultTypeBootstrap)
 	)
 
+	// Determine which peers own the specified shard
 	peers, err := s.peersForShard(shard)
 	if err != nil {
 		return nil, err
 	}
 
+	// Emit a gauge indicating whether we're done or not
 	go func() {
-		for atomic.LoadInt64(&complete) == 0 {
-			m.fetchBlocksFromPeers.Update(1)
-			time.Sleep(gaugeReportInterval)
+		for {
+			select {
+			case <-doneCh:
+				progress.fetchBlocksFromPeers.Update(0)
+				return
+			default:
+				progress.fetchBlocksFromPeers.Update(1)
+				time.Sleep(gaugeReportInterval)
+			}
 		}
-		m.fetchBlocksFromPeers.Update(0)
 	}()
+	defer close(doneCh)
 
 	// Begin pulling metadata, if one or multiple peers fail no error will
 	// be returned from this routine as long as one peer succeeds completely
-	metadataCh := make(chan blocksMetadata, 4096)
+	metadataCh := make(chan blocksMetadata, blocksMetadataChBufSize)
+	// Spin up a background goroutine which will begin streaming metadata from
+	// all the peers and pushing them into the metadatach
+	errCh := make(chan error, 1)
 	go func() {
-		err := s.streamBlocksMetadataFromPeers(nsMetadata.ID(), shard, peers,
-			start, end, metadataCh, m)
-
+		errCh <- s.streamBlocksMetadataFromPeers(
+			nsMetadata.ID(), shard, peers, start, end, metadataCh, progress, version)
 		close(metadataCh)
-		if err != nil {
-			// Bail early
-			onDone(err)
-		}
 	}()
 
-	// Begin consuming metadata and making requests
-	go func() {
-		err := s.streamBlocksFromPeers(nsMetadata, shard, peers,
-			metadataCh, opts, result, m, s.streamAndGroupCollectedBlocksMetadata)
-		onDone(err)
-	}()
+	// Begin consuming metadata and making requests. This will block until all
+	// data has been streamed (or failed to stream). Note that this function does
+	// not return an error and if anything goes wrong here we won't report it to
```

Comment:
Seems strange that error is not bubbled up here, not sure what's changed that we don't want to bubble up the error anymore - we used to have the behavior described here but allowed some very heavy errors to continue to bubble up.

Reply:
If you look at the source code for streamBlocksFromPeers on master, the function signature returns an error, but there is only a single call to return in that function in which it explicitly returns nil, so this was a minor refactoring where I updated the signature of that function since it was not possible for it to ever return an error

Reply:
I see, ok yeah makes sense.  If we want to add the possibility to float up an error later from that method we can just change this back.

Reply:
yeah exactly, better to have a useful function signature in the meantime I think